File: source/ports/scala_port/src/main/scala/Loader.scala
package metacall

import metacall.util._
import java.nio.file.Paths
import com.sun.jna._, ptr.PointerByReference
import scala.util._

/** Loads scripts into MetaCall
  * NOTE: Assumes MetaCall is initialized
  * WARNNING: You should always load using `Caller.loadFile` so all files are loaded
  * on MetaCall's thread.
  */
private[metacall] object Loader {

  /** NOTE: Should only be called from the MetaCall thread */
  private[metacall] def loadFilesUnsafe(
      runtime: Runtime,
      filePaths: Vector[String],
      handleRef: Option[PointerByReference]
  ): Try[Unit] = {
    val absolutePaths =
      filePaths.map(filePath => Paths.get(filePath).toAbsolutePath().toString())

    val code = Bindings.instance.metacall_load_from_file(
      runtime.toString(),
      absolutePaths.toArray,
      SizeT(absolutePaths.size.asInstanceOf[Long]),
      handleRef match {
        case Some(ref) => ref
        case None      => null
      }
    )

    if (code != 0)
      Failure(new Exception("Failed to load scripts: " + filePaths.mkString(" ")))
    else
      Success(())
  }
}

sealed trait Runtime
object Runtime {

  case object Python extends Runtime {
    override def toString() = "py"
  }

  case object Node extends Runtime {
    override def toString() = "node"
  }

}

File: source/ports/java_port/src/main/java/metacall/Loader.java
package metacall;

import java.nio.file.Paths;
import java.util.*;
import com.sun.jna.ptr.PointerByReference;

import metacall.util.*;

class Loader implements Runtime{
	int code;
	void loadFilesUnsafe(Runtime runtime, Vector<String> filePaths, Optional<PointerByReference> handleRef) throws Exception
	{
		//TODO
		// absolutePaths
		// code

		if (code != 0)
		{
			throw new Exception("Failed to load scripts" + Arrays.toString(filePaths.toArray()));
		}
		else
		{
			System.out.println("Success");
		}
	}
}

interface Runtime
{
	class Python
	{
		private Python() {}
		@Override
		public String toString() {
			return "py";
		}
	}

	class Node
	{
		private Node() {}
		@Override
		public String toString() {
			return "node";
		}
	}
}
